### Mô tả yêu cầu bài toán
1. Các hồ sơ (tuổi tác / nghề nghiệp / tình trạng hôn nhân / v.v.) nào có khả năng gửi tiền nhất?

2. Tình hình tài chính của mọi người có ảnh hưởng đến cách họ trả lời không?

3. Thời gian (ngày tháng / thời gian kéo dài) và loại hình liên hệ có quan trọng không? Làm thế nào? Có giống nhau cho tất cả các nhóm người không?

4. Số lần liên hệ tốt nhất để đạt được sự đồng tình là bao nhiêu? Khoảng thời gian tốt nhất là bao lâu? Người ta có khả năng đồng ý với chiến dịch nếu họ đã đồng ý trong quá khứ không?

5. Tình hình kinh tế (tỷ lệ biến động việc làm / Chỉ số Giá tiêu dùng / Chỉ số Tự tin của người tiêu dùng / Euribor / v.v.) có ảnh hưởng không? Làm thế nào? Có giống nhau cho tất cả mọi người không?

6. Có những thông tin thú vị nào để thêm vào không?

In [1]:
# Mount GDrive to load data
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
#import thư viện
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

### Load data

In [38]:
df_banking = pd.read_csv('/content/drive/MyDrive/baitL/bank-additional-full.csv')
df_banking.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   index,"age        41188 non-null  object 
 1   job""             41188 non-null  object 
 2   marital""         41188 non-null  object 
 3   education""       41188 non-null  object 
 4   default""         41188 non-null  object 
 5   housing""         41188 non-null  object 
 6   loan""            41188 non-null  object 
 7   contact""         41188 non-null  object 
 8   month""           41188 non-null  object 
 9   day_of_week""     41188 non-null  object 
 10  duration""        41188 non-null  int64  
 11  campaign""        41188 non-null  int64  
 12  pdays""           41188 non-null  int64  
 13  previous""        41188 non-null  int64  
 14  poutcome""        41188 non-null  object 
 15  emp.var.rate""    41188 non-null  float64
 16  cons.price.idx""  41188 non-null  float6

In [39]:
# hiển thị 10 cột đầu tiên
df_banking.head(10)

index,"age          job""  marital""            education""  default""  \
0      0,"56    housemaid""  married""             basic.4y""       no""   
1      1,"57     services""  married""          high.school""  unknown""   
2      2,"37     services""  married""          high.school""       no""   
3      3,"40       admin.""  married""             basic.6y""       no""   
4      4,"56     services""  married""          high.school""       no""   
5      5,"45     services""  married""             basic.9y""  unknown""   
6      6,"59       admin.""  married""  professional.course""       no""   
7      7,"41  blue-collar""  married""              unknown""  unknown""   
8      8,"24   technician""   single""  professional.course""       no""   
9      9,"25     services""   single""          high.school""       no""   

  housing"" loan""    contact"" month"" day_of_week""  ...  campaign""  \
0      no""   no""  telephone""   may""         mon""  ...           1   
1      no""   no""  telephone""   may""         mon""  ...           1   
2     yes""   no""  telephone""   may""         mon""  ...           1   
3      no""   no""  telephone""   may""         mon""  ...           1   
4      no""  yes""  telephone""   may""         mon""  ...           1   
5      no""   no""  telephone""   may""         mon""  ...           1   
6      no""   no""  telephone""   may""         mon""  ...           1   
7      no""   no""  telephone""   may""         mon""  ...           1   
8     yes""   no""  telephone""   may""         mon""  ...           1   
9     yes""   no""  telephone""   may""         mon""  ...           1   

   pdays""  previous""     poutcome"" emp.var.rate""  cons.price.idx""  \
0      999           0  nonexistent""            1.1            93.994   
1      999           0  nonexistent""            1.1            93.994   
2      999           0  nonexistent""            1.1            93.994   
3      999           0  nonexistent""            1.1            93.994   
4      999           0  nonexistent""            1.1            93.994   
5      999           0  nonexistent""            1.1            93.994   
6      999           0  nonexistent""            1.1            93.994   
7      999           0  nonexistent""            1.1            93.994   
8      999           0  nonexistent""            1.1            93.994   
9      999           0  nonexistent""            1.1            93.994   

   cons.conf.idx""  euribor3m""  nr.employed""     y"""  
0            -36.4        4.857         5191.0  no""\r"  
1            -36.4        4.857         5191.0  no""\r"  
2            -36.4        4.857         5191.0  no""\r"  
3            -36.4        4.857         5191.0  no""\r"  
4            -36.4        4.857         5191.0  no""\r"  
5            -36.4        4.857         5191.0  no""\r"  
6            -36.4        4.857         5191.0  no""\r"  
7            -36.4        4.857         5191.0  no""\r"  
8            -36.4        4.857         5191.0  no""\r"  
9            -36.4        4.857         5191.0  no""\r"  

[10 rows x 21 columns]

In [40]:

first_column_name = df_banking.columns[0]

# Đổi tên cột đầu tiên chỉ lấy 3 ký tự cuối
df_banking.rename(columns={first_column_name: first_column_name[-3:]}, inplace=True)
df_banking

age          job""  marital""            education""  default""  \
0          0,"56    housemaid""  married""             basic.4y""       no""   
1          1,"57     services""  married""          high.school""  unknown""   
2          2,"37     services""  married""          high.school""       no""   
3          3,"40       admin.""  married""             basic.6y""       no""   
4          4,"56     services""  married""          high.school""       no""   
...          ...            ...        ...                    ...        ...   
41183  41183,"73      retired""  married""  professional.course""       no""   
41184  41184,"46  blue-collar""  married""  professional.course""       no""   
41185  41185,"56      retired""  married""    university.degree""       no""   
41186  41186,"44   technician""  married""  professional.course""       no""   
41187  41187,"74      retired""  married""  professional.course""       no""   

      housing"" loan""    contact"" month"" day_of_week""  ...  campaign""  \
0          no""   no""  telephone""   may""         mon""  ...           1   
1          no""   no""  telephone""   may""         mon""  ...           1   
2         yes""   no""  telephone""   may""         mon""  ...           1   
3          no""   no""  telephone""   may""         mon""  ...           1   
4          no""  yes""  telephone""   may""         mon""  ...           1   
...         ...    ...          ...     ...           ...  ...         ...   
41183     yes""   no""   cellular""   nov""         fri""  ...           1   
41184      no""   no""   cellular""   nov""         fri""  ...           1   
41185     yes""   no""   cellular""   nov""         fri""  ...           2   
41186      no""   no""   cellular""   nov""         fri""  ...           1   
41187     yes""   no""   cellular""   nov""         fri""  ...           3   

       pdays""  previous""     poutcome"" emp.var.rate""  cons.price.idx""  \
0          999           0  nonexistent""            1.1            93.994   
1          999           0  nonexistent""            1.1            93.994   
2          999           0  nonexistent""            1.1            93.994   
3          999           0  nonexistent""            1.1            93.994   
4          999           0  nonexistent""            1.1            93.994   
...        ...         ...            ...            ...               ...   
41183      999           0  nonexistent""           -1.1            94.767   
41184      999           0  nonexistent""           -1.1            94.767   
41185      999           0  nonexistent""           -1.1            94.767   
41186      999           0  nonexistent""           -1.1            94.767   
41187      999           1      failure""           -1.1            94.767   

       cons.conf.idx""  euribor3m""  nr.employed""      y"""  
0                -36.4        4.857         5191.0   no""\r"  
1                -36.4        4.857         5191.0   no""\r"  
2                -36.4        4.857         5191.0   no""\r"  
3                -36.4        4.857         5191.0   no""\r"  
4                -36.4        4.857         5191.0   no""\r"  
...                ...          ...            ...       ...  
41183            -50.8        1.028         4963.6  yes""\r"  
41184            -50.8        1.028         4963.6   no""\r"  
41185            -50.8        1.028         4963.6   no""\r"  
41186            -50.8        1.028         4963.6  yes""\r"  
41187            -50.8        1.028         4963.6   no""\r"  

[41188 rows x 21 columns]

In [41]:
# Lấy 2 số cuối cùng của cột Age
df_banking["age"] = df_banking["age"].str[-2:]

df_banking.sample(10)

age          job""   marital""            education""  default""  \
35794  56   technician""    single""  professional.course""       no""   
16812  49  blue-collar""   married""             basic.9y""  unknown""   
25392  48   technician""    single""    university.degree""       no""   
1343   28  blue-collar""   married""             basic.6y""       no""   
27483  32       admin.""  divorced""    university.degree""       no""   
16686  48   technician""  divorced""  professional.course""       no""   
32893  30  blue-collar""   married""             basic.6y""       no""   
18841  58  blue-collar""   married""             basic.9y""       no""   
24233  41   technician""    single""  professional.course""       no""   
14600  30  blue-collar""    single""             basic.4y""       no""   

       housing""     loan""    contact"" month"" day_of_week""  ...  \
35794      yes""       no""   cellular""   may""         mon""  ...   
16812       no""       no""   cellular""   jul""         thu""  ...   
25392      yes""      yes""   cellular""   nov""         tue""  ...   
1343        no""       no""  telephone""   may""         thu""  ...   
27483      yes""       no""   cellular""   nov""         fri""  ...   
16686       no""       no""   cellular""   jul""         wed""  ...   
32893      yes""       no""   cellular""   may""         mon""  ...   
18841      yes""       no""   cellular""   aug""         mon""  ...   
24233  unknown""  unknown""   cellular""   nov""         mon""  ...   
14600      yes""       no""   cellular""   jul""         tue""  ...   

       campaign""  pdays""  previous""     poutcome"" emp.var.rate""  \
35794           3      999           1      failure""           -1.8   
16812           1      999           0  nonexistent""            1.4   
25392           4      999           1      failure""           -0.1   
1343            1      999           0  nonexistent""            1.1   
27483           3      999           0  nonexistent""           -0.1   
16686           2      999           0  nonexistent""            1.4   
32893           1      999           1      failure""           -1.8   
18841           1      999           0  nonexistent""            1.4   
24233           1      999           0  nonexistent""           -0.1   
14600           2      999           0  nonexistent""            1.4   

       cons.price.idx""  cons.conf.idx""  euribor3m""  nr.employed""     y"""  
35794            92.893            -46.2        1.244         5099.1  no""\r"  
16812            93.918            -42.7        4.962         5228.1  no""\r"  
25392            93.200            -42.0        4.153         5195.8  no""\r"  
1343             93.994            -36.4        4.855         5191.0  no""\r"  
27483            93.200            -42.0        4.021         5195.8  no""\r"  
16686            93.918            -42.7        4.963         5228.1  no""\r"  
32893            92.893            -46.2        1.299         5099.1  no""\r"  
18841            93.444            -36.1        4.970         5228.1  no""\r"  
24233            93.200            -42.0        4.191         5195.8  no""\r"  
14600            93.918            -42.7        4.961         5228.1  no""\r"  

[10 rows x 21 columns]

#### Thay Thế tên cột hàng loạt

In [42]:
# Lấy danh sách các tên cột từ cột thứ 2 trừ cột y
columns_to_rename = df_banking.columns[1:-1]

# Đổi tên các cột từ cột thứ 2 trở đi bỏ đi 2 ký tự cuối
df_banking.rename(columns={col: col[:-2] for col in columns_to_rename}, inplace=True)


# Lấy tên cột cuối cùng
last_column_name = df_banking.columns[-1]

# Đổi tên cột cuối cùng bỏ đi 3 ký tự cuối
df_banking.rename(columns={last_column_name: last_column_name[:-3]}, inplace=True)
df_banking

age            job    marital              education    default housing  \
0      56    housemaid""  married""             basic.4y""       no""    no""   
1      57     services""  married""          high.school""  unknown""    no""   
2      37     services""  married""          high.school""       no""   yes""   
3      40       admin.""  married""             basic.6y""       no""    no""   
4      56     services""  married""          high.school""       no""    no""   
...    ..            ...        ...                    ...        ...     ...   
41183  73      retired""  married""  professional.course""       no""   yes""   
41184  46  blue-collar""  married""  professional.course""       no""    no""   
41185  56      retired""  married""    university.degree""       no""   yes""   
41186  44   technician""  married""  professional.course""       no""    no""   
41187  74      retired""  married""  professional.course""       no""   yes""   

        loan      contact  month day_of_week  ...  campaign  pdays  previous  \
0       no""  telephone""  may""       mon""  ...         1    999         0   
1       no""  telephone""  may""       mon""  ...         1    999         0   
2       no""  telephone""  may""       mon""  ...         1    999         0   
3       no""  telephone""  may""       mon""  ...         1    999         0   
4      yes""  telephone""  may""       mon""  ...         1    999         0   
...      ...          ...    ...         ...  ...       ...    ...       ...   
41183   no""   cellular""  nov""       fri""  ...         1    999         0   
41184   no""   cellular""  nov""       fri""  ...         1    999         0   
41185   no""   cellular""  nov""       fri""  ...         2    999         0   
41186   no""   cellular""  nov""       fri""  ...         1    999         0   
41187   no""   cellular""  nov""       fri""  ...         3    999         1   

            poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent""          1.1          93.994          -36.4      4.857   
1      nonexistent""          1.1          93.994          -36.4      4.857   
2      nonexistent""          1.1          93.994          -36.4      4.857   
3      nonexistent""          1.1          93.994          -36.4      4.857   
4      nonexistent""          1.1          93.994          -36.4      4.857   
...              ...          ...             ...            ...        ...   
41183  nonexistent""         -1.1          94.767          -50.8      1.028   
41184  nonexistent""         -1.1          94.767          -50.8      1.028   
41185  nonexistent""         -1.1          94.767          -50.8      1.028   
41186  nonexistent""         -1.1          94.767          -50.8      1.028   
41187      failure""         -1.1          94.767          -50.8      1.028   

       nr.employed         y  
0           5191.0   no""\r"  
1           5191.0   no""\r"  
2           5191.0   no""\r"  
3           5191.0   no""\r"  
4           5191.0   no""\r"  
...            ...       ...  
41183       4963.6  yes""\r"  
41184       4963.6   no""\r"  
41185       4963.6   no""\r"  
41186       4963.6  yes""\r"  
41187       4963.6   no""\r"  

[41188 rows x 21 columns]

In [43]:
#bỏ các dấu "" trong data
df_banking = df_banking.applymap(lambda x: x.replace('"', '') if isinstance(x, str) else x)
df_banking

age          job  marital            education  default housing loan  \
0      56    housemaid  married             basic.4y       no      no   no   
1      57     services  married          high.school  unknown      no   no   
2      37     services  married          high.school       no     yes   no   
3      40       admin.  married             basic.6y       no      no   no   
4      56     services  married          high.school       no      no  yes   
...    ..          ...      ...                  ...      ...     ...  ...   
41183  73      retired  married  professional.course       no     yes   no   
41184  46  blue-collar  married  professional.course       no      no   no   
41185  56      retired  married    university.degree       no     yes   no   
41186  44   technician  married  professional.course       no      no   no   
41187  74      retired  married  professional.course       no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
0      telephone   may         mon  ...         1    999         0   
1      telephone   may         mon  ...         1    999         0   
2      telephone   may         mon  ...         1    999         0   
3      telephone   may         mon  ...         1    999         0   
4      telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
41183   cellular   nov         fri  ...         1    999         0   
41184   cellular   nov         fri  ...         1    999         0   
41185   cellular   nov         fri  ...         2    999         0   
41186   cellular   nov         fri  ...         1    999         0   
41187   cellular   nov         fri  ...         3    999         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent          1.1          93.994          -36.4      4.857   
1      nonexistent          1.1          93.994          -36.4      4.857   
2      nonexistent          1.1          93.994          -36.4      4.857   
3      nonexistent          1.1          93.994          -36.4      4.857   
4      nonexistent          1.1          93.994          -36.4      4.857   
...            ...          ...             ...            ...        ...   
41183  nonexistent         -1.1          94.767          -50.8      1.028   
41184  nonexistent         -1.1          94.767          -50.8      1.028   
41185  nonexistent         -1.1          94.767          -50.8      1.028   
41186  nonexistent         -1.1          94.767          -50.8      1.028   
41187      failure         -1.1          94.767          -50.8      1.028   

       nr.employed      y  
0           5191.0   no\r  
1           5191.0   no\r  
2           5191.0   no\r  
3           5191.0   no\r  
4           5191.0   no\r  
...            ...    ...  
41183       4963.6  yes\r  
41184       4963.6   no\r  
41185       4963.6   no\r  
41186       4963.6  yes\r  
41187       4963.6   no\r  

[41188 rows x 21 columns]

In [44]:
# Xóa ký tự cuối của cột 'y'
df_banking['y'] = df_banking['y'].apply(lambda x: x[:-1])

In [45]:
df_banking

age          job  marital            education  default housing loan  \
0      56    housemaid  married             basic.4y       no      no   no   
1      57     services  married          high.school  unknown      no   no   
2      37     services  married          high.school       no     yes   no   
3      40       admin.  married             basic.6y       no      no   no   
4      56     services  married          high.school       no      no  yes   
...    ..          ...      ...                  ...      ...     ...  ...   
41183  73      retired  married  professional.course       no     yes   no   
41184  46  blue-collar  married  professional.course       no      no   no   
41185  56      retired  married    university.degree       no     yes   no   
41186  44   technician  married  professional.course       no      no   no   
41187  74      retired  married  professional.course       no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
0      telephone   may         mon  ...         1    999         0   
1      telephone   may         mon  ...         1    999         0   
2      telephone   may         mon  ...         1    999         0   
3      telephone   may         mon  ...         1    999         0   
4      telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
41183   cellular   nov         fri  ...         1    999         0   
41184   cellular   nov         fri  ...         1    999         0   
41185   cellular   nov         fri  ...         2    999         0   
41186   cellular   nov         fri  ...         1    999         0   
41187   cellular   nov         fri  ...         3    999         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent          1.1          93.994          -36.4      4.857   
1      nonexistent          1.1          93.994          -36.4      4.857   
2      nonexistent          1.1          93.994          -36.4      4.857   
3      nonexistent          1.1          93.994          -36.4      4.857   
4      nonexistent          1.1          93.994          -36.4      4.857   
...            ...          ...             ...            ...        ...   
41183  nonexistent         -1.1          94.767          -50.8      1.028   
41184  nonexistent         -1.1          94.767          -50.8      1.028   
41185  nonexistent         -1.1          94.767          -50.8      1.028   
41186  nonexistent         -1.1          94.767          -50.8      1.028   
41187      failure         -1.1          94.767          -50.8      1.028   

       nr.employed    y  
0           5191.0   no  
1           5191.0   no  
2           5191.0   no  
3           5191.0   no  
4           5191.0   no  
...            ...  ...  
41183       4963.6  yes  
41184       4963.6   no  
41185       4963.6   no  
41186       4963.6  yes  
41187       4963.6   no  

[41188 rows x 21 columns]

In [46]:
# đưa kiểu dữ liệu của cột Age từ object thành int
df_banking['age'] = df_banking['age'].astype(int)
df_banking.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

In [47]:
#lưu file kết quả chuẩn bị dữ liệu:
df_banking.to_csv('/content/drive/MyDrive/baitL/bank-additional-full-processed.csv', index=None)
print('đã lưu')

đã lưu


### Load data đã xử lý

In [98]:
df_bank_processed = pd.read_csv('/content/drive/MyDrive/baitL/bank-additional-full-processed.csv')
df_bank_processed

age          job  marital            education  default housing loan  \
0       56    housemaid  married             basic.4y       no      no   no   
1       57     services  married          high.school  unknown      no   no   
2       37     services  married          high.school       no     yes   no   
3       40       admin.  married             basic.6y       no      no   no   
4       56     services  married          high.school       no      no  yes   
...    ...          ...      ...                  ...      ...     ...  ...   
41183   73      retired  married  professional.course       no     yes   no   
41184   46  blue-collar  married  professional.course       no      no   no   
41185   56      retired  married    university.degree       no     yes   no   
41186   44   technician  married  professional.course       no      no   no   
41187   74      retired  married  professional.course       no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
0      telephone   may         mon  ...         1    999         0   
1      telephone   may         mon  ...         1    999         0   
2      telephone   may         mon  ...         1    999         0   
3      telephone   may         mon  ...         1    999         0   
4      telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
41183   cellular   nov         fri  ...         1    999         0   
41184   cellular   nov         fri  ...         1    999         0   
41185   cellular   nov         fri  ...         2    999         0   
41186   cellular   nov         fri  ...         1    999         0   
41187   cellular   nov         fri  ...         3    999         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent          1.1          93.994          -36.4      4.857   
1      nonexistent          1.1          93.994          -36.4      4.857   
2      nonexistent          1.1          93.994          -36.4      4.857   
3      nonexistent          1.1          93.994          -36.4      4.857   
4      nonexistent          1.1          93.994          -36.4      4.857   
...            ...          ...             ...            ...        ...   
41183  nonexistent         -1.1          94.767          -50.8      1.028   
41184  nonexistent         -1.1          94.767          -50.8      1.028   
41185  nonexistent         -1.1          94.767          -50.8      1.028   
41186  nonexistent         -1.1          94.767          -50.8      1.028   
41187      failure         -1.1          94.767          -50.8      1.028   

       nr.employed    y  
0           5191.0   no  
1           5191.0   no  
2           5191.0   no  
3           5191.0   no  
4           5191.0   no  
...            ...  ...  
41183       4963.6  yes  
41184       4963.6   no  
41185       4963.6   no  
41186       4963.6  yes  
41187       4963.6   no  

[41188 rows x 21 columns]

In [99]:
# hiển thị 5 bản ghi đầu tiên
df_bank_processed.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

In [100]:
# chia ra 5 nhóm tuổi
df_bank_processed.loc[(df_bank_processed['age'] > 15) & (df_bank_processed['age'] <= 30), 'age'] = 1
df_bank_processed.loc[(df_bank_processed['age'] > 30) & (df_bank_processed['age'] <= 40), 'age'] = 2
df_bank_processed.loc[(df_bank_processed['age'] > 40) & (df_bank_processed['age'] <= 50), 'age'] = 3
df_bank_processed.loc[(df_bank_processed['age'] > 50) & (df_bank_processed['age'] <= 60), 'age'] = 4
df_bank_processed.loc[ df_bank_processed['age'] > 60, 'age'] = 5


### 1. What profiles ( age / job / marital status / etc.) are the most likely to respond positively ?

In [101]:
grouped_data = df_bank_processed.groupby('y')

def find_max_response_value(series):
    response_rates = series.value_counts(normalize=True)
    max_response_value = response_rates.idxmax()
    return max_response_value

max_response_age = grouped_data['age'].apply(find_max_response_value).loc['yes']
print("Độ tuổi gửi tiền vào ngân hàng nhiều nhất thuộc nhóm:", max_response_age)
max_response_job = grouped_data['job'].apply(find_max_response_value).loc['yes']
print("công việc gửi tiền vào ngân hàng nhiều nhất là:", max_response_job)
max_response_marital = grouped_data['marital'].apply(find_max_response_value).loc['yes']
print("người có tình trạng hôn nhân gửi tiền vào ngân hàng nhiều nhất thuộc nhóm:", max_response_marital)
max_response_education = grouped_data['education'].apply(find_max_response_value).loc['yes']
print("người có tình trạng giáo dục gửi tiền vào ngân hàng nhiều nhất thuộc nhóm:", max_response_education)


Độ tuổi gửi tiền vào ngân hàng nhiều nhất thuộc nhóm: 2
công việc gửi tiền vào ngân hàng nhiều nhất là: admin.
người có tình trạng hôn nhân gửi tiền vào ngân hàng nhiều nhất thuộc nhóm: married
người có tình trạng giáo dục gửi tiền vào ngân hàng nhiều nhất thuộc nhóm: university.degree


### 2. Does the financial situation of people have an impact on how they answer ?

In [112]:
# Chuyển đổi giá trị 'yes' thành 1 và 'no' thành 0 trong cột 'Response'
df_bank_processed['y'] = df_bank_processed['y'].map({'yes': 1, 'no': 0})
feedback_by_default = df_bank_processed.groupby('default')['y'].mean()
feedback_by_housing = df_bank_processed.groupby('housing')['y'].mean()
feedback_by_loan = df_bank_processed.groupby('loan')['y'].mean()

print("Tỷ lệ gửi tiền theo cột 'default':")
print(feedback_by_default.round(3))

print("Tỷ lệ gửi tiền theo cột 'housing':")
print(feedback_by_housing.round(3))

print("Tỷ lệ gửi tiền theo cột 'loan':")
print(feedback_by_loan.round(3))


Tỷ lệ gửi tiền theo cột 'default':
default
no         0.129
unknown    0.052
yes        0.000
Name: y, dtype: float64
Tỷ lệ gửi tiền theo cột 'housing':
housing
no         0.109
unknown    0.108
yes        0.116
Name: y, dtype: float64
Tỷ lệ gửi tiền theo cột 'loan':
loan
no         0.113
unknown    0.108
yes        0.109
Name: y, dtype: float64


### 3. Do the timing (date / duration) and type of contact matter significantly ?

In [114]:

# Tạo danh sách các nhóm bạn muốn phân tích
time_month = ['month']
time_day_of_week = [ 'day_of_week']
contact_groups = ['contact']


response_rates_time = {}
response_rates_duration = {}
response_rates_contact = {}

# Tính tỷ lệ phản hồi cho mỗi nhóm thời gian
for group in time_groups:
    grouped_data = df_bank_processed.groupby(group)
    response_rates_time[group] = grouped_data['y'].value_counts(normalize=True).unstack().fillna(0)


# Tính tỷ lệ phản hồi cho mỗi loại liên lạc (contact)
for group in contact_groups:
    grouped_data = df_bank_processed.groupby(group)
    response_rates_contact[group] = grouped_data['y'].value_counts(normalize=True).unstack().fillna(0)

# In kết quả
for group, rates in response_rates_time.items():
    print(f'Tỷ lệ phản hồi theo {group}:')
    print(rates.round(3))

for group, rates in response_rates_duration.items():
    print(f'Tỷ lệ phản hồi theo {group}:')
    print(rates.round(3))

for group, rates in response_rates_contact.items():
    print(f'Tỷ lệ phản hồi theo {group}:')
    print(rates.round(3))


Tỷ lệ phản hồi theo month:
y          0      1
month              
apr    0.795  0.205
aug    0.894  0.106
dec    0.511  0.489
jul    0.910  0.090
jun    0.895  0.105
mar    0.495  0.505
may    0.936  0.064
nov    0.899  0.101
oct    0.561  0.439
sep    0.551  0.449
Tỷ lệ phản hồi theo day_of_week:
y                0      1
day_of_week              
fri          0.892  0.108
mon          0.901  0.099
thu          0.879  0.121
tue          0.882  0.118
wed          0.883  0.117
Tỷ lệ phản hồi theo contact:
y              0      1
contact                
cellular   0.853  0.147
telephone  0.948  0.052


### 4 - What is the best number of contacts to reach agreement ?
### What is the best delay ?
### Are people more likely to agree to the campaign if they have already agreed in the past ?

In [119]:

best_campaign = df_bank_processed[df_bank_processed['y'] == 1]['campaign'].idxmin()
best_campaign_value = df_bank_processed.loc[best_campaign, 'campaign']

agreed_before = df_bank_processed[df_bank_processed['previous'] > 0]
agreed_before_rate = agreed_before['y'].value_counts(normalize=True).get(1, 0)

#Lọc ra các dòng có pdays khác -1 (đã được liên hệ trước đó)
filtered_data = df_bank_additional[df_bank_additional['pdays'] != -1]
average_delay_to_agree = filtered_data.groupby('pdays')['y'].mean()
best_delay = average_delay_to_agree.idxmax()

print("Khoảng thời gian tối ưu giữa các cuộc gọi tiếp thị:",best_delay, "ngày")

print(f"Số lần liên hệ tốt nhất để có thỏa thuận: {best_campaign_value}")

print(f"Tỷ lệ đạt thỏa thuận khi đã có thỏa thuận trước đây: {agreed_before_rate:.2%}")


Khoảng thời gian tối ưu giữa các cuộc gọi tiếp thị: 21 ngày
Số lần liên hệ tốt nhất để có thỏa thuận: 1
Tỷ lệ đạt thỏa thuận khi đã có thỏa thuận trước đây: 26.65%


#### nhận xét:
---
- số lần liên hệ tốt nhất là 1 lần sẽ có tỷ lệ phản hồi cao nhất.
- tốt nhất nên liên hệ lại sau 21 ngày.
- người đã đồng ý với chiến dịch trong quá khứ sẽ có tỷ lệ đồng ý ở chiến dịch mới cao hơn.

### 5. Does the economy (Employment variation rate / Consumer Price Index  / Consumer Confidence Index / Euribor / etc.) have an impact ? How ? Is it the same for everybody ?

In [ ]:
feedback_by_emp = df_bank_additional.groupby('emp.var.rate')['y'].mean()

print("\nTỷ lệ gửi tiền theo Employment variation rate")
print(feedback_by_emp)

feedback_by_cons_price = df_bank_additional.groupby('cons.price.idx')['y'].mean()

print("\nTỷ lệ gửi tiền theo Consumer Price Index")
print(feedback_by_cons_price)

feedback_by_cons_conf = df_bank_additional.groupby('cons.conf.idx')['y'].mean()

print("\nTỷ lệ gửi tiền theo Consumer Price Index")
print(feedback_by_cons_conf)


Tỷ lệ gửi tiền theo Employment variation rate
emp.var.rate
-3.4    42.390289
-3.0    51.162791
-2.9    35.718581
-1.8    15.908101
-1.7    52.134541
-1.1    47.401575
-0.2    10.000000
-0.1     6.299213
 1.1     3.091588
 1.4     5.334483
Name: y, dtype: float64

Tỷ lệ gửi tiền theo Consumer Price Index
cons.price.idx
92.201    34.285714
92.379    39.700375
92.431    40.268456
92.469    37.078652
92.649    47.058824
92.713    51.162791
92.756    10.000000
92.843    44.680851
92.893     9.043838
92.963    36.923077
93.075    17.982099
93.200     5.254425
93.369    56.818182
93.444     5.236715
93.749    55.747126
93.798    62.686567
93.876    57.547170
93.918     6.088257
93.994     3.091588
94.027    51.502146
94.055    46.724891
94.199    49.504950
94.215    56.591640
94.465     4.298125
94.601    45.588235
94.767    45.312500
Name: y, dtype: float64

Tỷ lệ gửi tiền theo Consumer Price Index
cons.conf.idx
-50.8    45.312500
-50.0    44.680851
-49.5    45.588235
-47.1    17.982099
-46

#### Nhận xét
1. Tỷ lệ gửi tiền theo Employment Variation Rate (emp.var.rate):

    - Có một biểu đồ tỷ lệ gửi tiền đối với Employment Variation Rate. Tỷ lệ phản hồi cao nhất xảy ra khi Employment Variation Rate ở mức -3.0 và -1.7, với tỷ lệ phản hồi lần lượt là 51.16% và 52.13%. Điều này ngụ ý rằng khi tỷ lệ biến đổi việc làm (Employment Variation Rate) thấp hoặc âm (có giảm sút), tỷ lệ gửi tiền cao hơn.

    - Khi Employment Variation Rate tăng lên (ví dụ: 1.4), tỷ lệ gửi tiền giảm xuống 5.33%. Điều này cho thấy một sự tương quan ngược giữa Employment Variation Rate dương và tỷ lệ gửi tiền.

2. Tỷ lệ gửi tiền theo Consumer Price Index (cons.price.idx):

    - Tỷ lệ gửi tiền dao động dọc theo Consumer Price Index. Tỷ lệ gửi tiền cao nhất xuất hiện khi Consumer Price Index nằm trong khoảng từ 93.369 đến 93.798, với tỷ lệ phản hồi từ 55.74% đến 62.69%. Điều này ngụ ý rằng khi chỉ số giá tiêu dùng (Consumer Price Index) ở mức này, tỷ lệ gửi tiền là cao nhất.

    - Khi Consumer Price Index tăng hoặc giảm khá xa khỏi khoảng giá trị này, tỷ lệ gửi tiền giảm xuống (ví dụ: 92.201 hoặc 94.027), chỉ còn khoảng 3-5%. Điều này cho thấy rằng có một mối quan hệ tương quan giữa Consumer Price Index và tỷ lệ gửi tiền.

3. Tỷ lệ gửi tiền theo Consumer Confidence Index (cons.conf.idx):

    - Tỷ lệ gửi tiền cũng thay đổi dựa trên Consumer Confidence Index. Tỷ lệ gửi tiền cao nhất xuất hiện khi Consumer Confidence Index nằm trong khoảng từ -40.0 đến -38.3, với tỷ lệ phản hồi từ 51.50% đến 62.69%. Điều này ngụ ý rằng khi chỉ số niềm tin của người tiêu dùng (Consumer Confidence Index) ở mức này, tỷ lệ gửi tiền là cao nhất.

    - Khi Consumer Confidence Index giảm thấp hơn hoặc tăng cao hơn, tỷ lệ gửi tiền giảm xuống (ví dụ: -36.4 hoặc -26.9), chỉ còn khoảng 3-5%. Điều này cho thấy rằng có một mối quan hệ tương quan giữa Consumer Confidence Index và tỷ lệ gửi tiền.

### 6. Any interesting insights to add ?

In [ ]:
# Tạo một DataFrame để lưu trữ thông kê chung
summary_data = pd.DataFrame(columns=['Column', 'Top Response', 'Response Rate'])

# Lặp qua từng cột trong DataFrame và xác định giá trị nổi trội có tỷ lệ phản hồi cao nhất
for column in df_bank_additional.columns:
    if df_bank_additional[column].dtype == 'object':  # Chỉ xem xét các cột có kiểu dữ liệu là object (categorical)
        top_response = df_bank_additional[column].value_counts().idxmax()
        response_rate = df_bank_additional[df_bank_additional[column] == top_response]['y'].mean()
        summary_data = pd.concat([summary_data, pd.DataFrame({'Column': [column], 'Top Response': [top_response], 'Response Rate': [response_rate]})], ignore_index=True)

# Hiển thị thông kê chung
print(summary_data)

      Column       Top Response Response Rate
0        job             admin.      0.129726
1    marital            married      0.101573
2  education  university.degree      0.137245
3    default                 no       0.12879
4    housing                yes      0.116194
5       loan                 no      0.113402
6    contact           cellular      0.147376
7      month                may      0.064347
8  dayofweek                thu      0.121188
9   poutcome        nonexistent      0.088322


#### Nhận xét

- Từ kết quả này, chúng ta có thể đưa ra một số nhận xét và kết luận sau:

    * Về mặt nhân khẩu học, những người có học vấn cao, đã kết hôn và đang làm việc trong lĩnh vực hành chính có xu hướng trả lời khảo sát cao hơn.
    * Về mặt tài chính, những người không có nợ xấu, có nhà ở và không có khoản vay có xu hướng trả lời khảo sát cao hơn.
    * Về mặt tiếp thị, những người được liên hệ qua điện thoại di động có xu hướng trả lời khảo sát cao hơn.

In [ ]:
# 1 - What profiles ( age / job / marital status / etc.) are the most likely to respond positively ?
# 2 - Does the financial situation of people have an impact on how they answer ?
# 3 - Do the timing (date / duration) and type of contact matter significantly ?
#    How ? Is it the same for all groups of people ?
# 4 - What is the best number of contacts to reach agreement ? What is the best delay ? Are people more likely to agree to the campaign if they have already agreed in the past ?
# 5 - Does the economy (Employment variation rate / Consumer Price Index  / Consumer Confidence Index / Euribor / etc.) have an impact ? How ? Is it the same for everybody ?
# 6 - Any interesting insights to add ?